src/mpid/common/shm/mpidu_shm_alloc.c

```diff
@@ -0,0 +1,702 @@
+                MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
+
+                mpl_err =
+                    MPL_shm_hnd_deserialize(shm_seg->hnd, serialized_hnd, strlen(serialized_hnd));
+                if (mpl_err) {
+                    *map_result_ptr = SYMSHM_OTHER_FAIL;
+                    goto result_sync;
+                }
+
+                mpl_err = MPL_shm_fixed_seg_attach(shm_seg->hnd, shm_seg->segment_len,
+                                                   (void **) &(shm_seg->base_addr), 0);
+                if (mpl_err) {
+                    *map_result_ptr =
+                        (mpl_err == MPL_SHM_EINVAL) ? SYMSHM_MAP_FAIL : SYMSHM_OTHER_FAIL;
+                } else
+                    mapped_flag = true;
+
+              result_sync:
+                /* check results of all processes. If any failure happens (max result > 0),
+                 * return SYMSHM_OTHER_FAIL if anyone reports it (max result == 2).
+                 * Otherwise return SYMSHM_MAP_FAIL (max result == 1). */
+                mpi_errno = MPIR_Allreduce(map_result_ptr, &all_map_result, 1, MPI_INT,
+                                           MPI_MAX, shm_comm_ptr, &errflag);
+                MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
+
+                if (all_map_result != SYMSHM_SUCCESS)
+                    goto map_fail;
+            }
+
+            mpi_errno = MPIR_Barrier(shm_comm_ptr, &errflag);
+            MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
+
+            if (shm_comm_ptr->rank == 0) {
+                /* shm_seg->hnd no longer needed */
+                mpl_err = MPL_shm_seg_remove(shm_seg->hnd);
+                MPIR_ERR_CHKANDJUMP(mpl_err, mpi_errno, MPI_ERR_OTHER, "**remove_shar_mem");
+            }
+        } else {
+            /* if there is only one process on this processor, don't use shared memory */
+            int rc = check_maprange_ok(shm_seg->base_addr, shm_seg->segment_len);
+            if (rc) {
+                shm_seg->base_addr = MPL_mmap(shm_seg->base_addr, shm_seg->segment_len,
+                                              PROT_READ | PROT_WRITE,
+                                              MAP_PRIVATE | MAP_ANON | MAP_FIXED, -1, 0,
+                                              MPL_MEM_SHM);
+            } else {
+                shm_seg->base_addr = (void *) MAP_FAILED;
+            }
+            *map_result_ptr = (shm_seg->base_addr == (void *) MAP_FAILED) ?
+                SYMSHM_MAP_FAIL : SYMSHM_SUCCESS;
+        }
+    }
+
+  fn_exit:
+    return mpi_errno;
+  map_fail:
+    /* clean up shm mapping resource */
+    if (mapped_flag) {
+        mpl_err =
+            MPL_shm_seg_detach(shm_seg->hnd, (void **) &(shm_seg->base_addr), shm_seg->segment_len);
+        MPIR_ERR_CHKANDJUMP(mpl_err, mpi_errno, MPI_ERR_OTHER, "**detach_shar_mem");
+    }
+    *map_result_ptr = all_map_result;
```

Comment:
I do not understand why you call `MPL_shm_seg_detach` and `MPL_shm_hnd_finalize` in separate places. Same question for `unmap_symm_shm` and `shm_alloc`. Is there any benefit to call `MPL_shm_hnd_finalize` in the caller functions?

Reply:
Reason is segments are initialized and finalized in the top functions, i.e., `MPIDU_shm_alloc_symm_all` and `MPIDU_shm_alloc`. Segment allocation and deallocation (`create_and_attach/attach` and `destroy`) happens one level down, i.e., in `shm_alloc_symm_all` and `shm_alloc`. The only reason for separating the two is readability of the code (IMO).

Reply:
@minsii are you happy with this or should I change it?

Reply:
I am fine with it.